pkg/apis/core/validation/validation.go

```diff
@@ -421,6 +421,34 @@ func ValidateVolumes(volumes []core.Volume, podMeta *metav1.ObjectMeta, fldPath
 	return vols, allErrs
 }
 
+func ValidateVolumesUpdate(volumes, oldVolumes []core.Volume, fldPath *field.Path, opts PodValidationOptions) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	mapOldVolumeNames := make(map[string]core.Volume)
+	for _, oldVol := range oldVolumes {
+		o := oldVol.DeepCopy()
+		if o.VolumeSource.Ephemeral == nil {
+			continue
+		}
+		mapOldVolumeNames[o.Name] = *o
+	}
+	for i, vol := range volumes {
+		oldVol, ok := mapOldVolumeNames[vol.Name]
+		idxPath := fldPath.Index(i)
+		if !ok {
+			allErrs = append(allErrs, validateVolumeSource(&vol.VolumeSource, idxPath, vol.Name, nil, opts)...)
+			continue
+		}
+		// If the old volume has limits set, then we will skip the validation for the volume, but verify that
+		// for any old volume that does not have a limit.
+		if _, ok := oldVol.VolumeSource.Ephemeral.VolumeClaimTemplate.Spec.Resources.Limits[core.ResourceStorage]; ok {
```

Comment:
Why is this check necessary? validateVolumeSourceUpdate covers this case AFAICT (my concern is that duplicating logic will be hard to maintain).

Reply:
I agree, duplicating logic is hard to maintain. Fixed.